File: backend-nodejs/src/models/Appointment.js
const { v4: uuidv4 } = require('uuid');
const { body, validationResult } = require('express-validator');

class Appointment {
  constructor(data = {}) {
    this.id = data.id || uuidv4();
    this.customer_id = data.customer_id || '';
    this.staff_id = data.staff_id || '';
    this.service_id = data.service_id || '';
    this.appointment_date = data.appointment_date || new Date();
    this.status = data.status || 'SCHEDULED';
    this.notes = data.notes || '';
    this.created_at = data.created_at || new Date();
    this.updated_at = data.updated_at || new Date();
  }

  // Static method to create validation rules
  static getValidationRules() {
    return [
      body('customer_id')
        .notEmpty()
        .withMessage('ID do cliente é obrigatório')
        .isUUID()
        .withMessage('ID do cliente deve ser um UUID válido'),
      
      body('staff_id')
        .notEmpty()
        .withMessage('ID do funcionário é obrigatório')
        .isUUID()
        .withMessage('ID do funcionário deve ser um UUID válido'),
      
      body('service_id')
        .notEmpty()
        .withMessage('ID do serviço é obrigatório')
        .isUUID()
        .withMessage('ID do serviço deve ser um UUID válido'),
      
      body('appointment_date')
        .notEmpty()
        .withMessage('Data do agendamento é obrigatória')
        .isISO8601()
        .withMessage('Data deve estar no formato ISO 8601')
        .custom((value) => {
          const appointmentDate = new Date(value);
          const now = new Date();
          if (appointmentDate <= now) {
            throw new Error('Data do agendamento deve ser no futuro');
          }
          return true;
        }),
      
      body('status')
        .optional()
        .isIn(['SCHEDULED', 'CONFIRMED', 'IN_PROGRESS', 'COMPLETED', 'CANCELLED', 'NO_SHOW'])
        .withMessage('Status deve ser um dos valores válidos: SCHEDULED, CONFIRMED, IN_PROGRESS, COMPLETED, CANCELLED, NO_SHOW'),
      
      body('notes')
        .optional()
        .isLength({ max: 500 })
        .withMessage('Observações devem ter no máximo 500 caracteres')
    ];
  }

  // Static method to validate request
  static validateRequest(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'Validation failed',
        details: errors.array(),
        timestamp: new Date().toISOString()
      });
    }
    next();
  }

  // Update method
  update(data) {
    if (data.customer_id !== undefined) this.customer_id = data.customer_id;
    if (data.staff_id !== undefined) this.staff_id = data.staff_id;
    if (data.service_id !== undefined) this.service_id = data.service_id;
    if (data.appointment_date !== undefined) this.appointment_date = new Date(data.appointment_date);
    if (data.status !== undefined) this.status = data.status;
    if (data.notes !== undefined) this.notes = data.notes;
    this.updated_at = new Date();
  }

  // Convert to plain object for database storage
  toPlainObject() {
    return {
      id: this.id,
      customer_id: this.customer_id,
      staff_id: this.staff_id,
      service_id: this.service_id,
      appointment_date: this.appointment_date,
      status: this.status,
      notes: this.notes,
      created_at: this.created_at,
      updated_at: this.updated_at
    };
  }

  // Create from database row
  static fromRow(row) {
    return new Appointment({
      id: row.id,
      customer_id: row.customer_id,
      staff_id: row.staff_id,
      service_id: row.service_id,
      appointment_date: row.appointment_date,
      status: row.status,
      notes: row.notes,
      created_at: row.created_at,
      updated_at: row.updated_at
    });
  }

  // Static method to get valid status values
  static getValidStatuses() {
    return ['SCHEDULED', 'CONFIRMED', 'IN_PROGRESS', 'COMPLETED', 'CANCELLED', 'NO_SHOW'];
  }
}

module.exports = Appointment;

File: backend/java-reactive/src/main/java/com/beautysalon/reactive/model/Appointment.java
package com.beautysalon.reactive.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.cassandra.core.mapping.PrimaryKey;
import org.springframework.data.cassandra.core.mapping.Table;

import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;
import java.util.UUID;

@Table("appointments")
public record Appointment(
    @PrimaryKey @Id UUID id,
    @NotNull(message = "Customer ID is required") UUID customerId,
    @NotNull(message = "Service ID is required") UUID serviceId,
    @NotNull(message = "Staff ID is required") UUID staffId,
    @NotNull(message = "Appointment date is required") LocalDateTime appointmentDate,
    String status,
    String notes,
    LocalDateTime createdAt,
    LocalDateTime updatedAt
) {
    public Appointment {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (status == null) {
            status = "SCHEDULED";
        }
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (updatedAt == null) {
            updatedAt = LocalDateTime.now();
        }
    }

    public static Appointment create(UUID customerId, UUID serviceId, UUID staffId, 
                                   LocalDateTime appointmentDate, String notes) {
        return new Appointment(null, customerId, serviceId, staffId, appointmentDate, null, notes, null, null);
    }

    public Appointment withUpdatedFields(LocalDateTime appointmentDate, String status, String notes) {
        return new Appointment(
            this.id,
            this.customerId,
            this.serviceId,
            this.staffId,
            appointmentDate != null ? appointmentDate : this.appointmentDate,
            status != null ? status : this.status,
            notes != null ? notes : this.notes,
            this.createdAt,
            LocalDateTime.now()
        );
    }

    public enum Status {
        SCHEDULED, CONFIRMED, IN_PROGRESS, COMPLETED, CANCELLED, NO_SHOW
    }
}
